src/CloudNative.CloudEvents.AspNetCore/HttpRequestExtension.cs

```diff
@@ -45,90 +22,54 @@ public static class HttpRequestExtension
-                    }
-                }
-
-                return await formatter.DecodeStructuredEventAsync(httpRequest.Body, extensions);
+                // TODO: Handle formatter being null
+                return await formatter.DecodeStructuredEventAsync(httpRequest.Body, extensionAttributes).ConfigureAwait(false);
             }
             else
             {
+                var headers = httpRequest.Headers;
                 CloudEventsSpecVersion version = CloudEventsSpecVersion.Default;
-                if (httpRequest.Headers[SpecVersionHttpHeader1] != StringValues.Empty)
+                if (headers.TryGetValue(HttpUtilities.SpecVersionHttpHeader, out var values))
                 {
-                    version = CloudEventsSpecVersion.V0_1;
+                    string versionId = values.First();
+                    version = CloudEventsSpecVersion.FromVersionId(versionId);
                 }
 
-                if (httpRequest.Headers[SpecVersionHttpHeader2] != StringValues.Empty)
+                var cloudEvent = new CloudEvent(version, extensionAttributes);
+                foreach (var header in headers)
                 {
-                    switch (httpRequest.Headers[SpecVersionHttpHeader2])
+                    string attributeName = HttpUtilities.GetAttributeNameFromHeaderName(header.Key);
+                    if (attributeName is null || attributeName == CloudEventsSpecVersion.SpecVersionAttribute.Name)
                     {
-                        case "0.2":
-                            version = CloudEventsSpecVersion.V0_2;
-                            break;
-                        case "0.3":
-                            version = CloudEventsSpecVersion.V0_3;
-                            break;
-                        default:
-                            version = CloudEventsSpecVersion.Default;
-                            break;
+                        continue;
                     }
+                    string attributeValue = HttpUtilities.DecodeHeaderValue(header.Value.First());
+
+                    cloudEvent.SetAttributeFromString(attributeName, attributeValue);
                 }
 
-                var cloudEvent = new CloudEvent(version, extensions);
-                var attributes = cloudEvent.GetAttributes();
-                foreach (var httpRequestHeader in httpRequest.Headers.Keys)
+                cloudEvent.DataContentType = httpRequest.ContentType;
+                if (httpRequest.Body is Stream body)
                 {
-                    if (httpRequestHeader.Equals(SpecVersionHttpHeader1,
-                            StringComparison.InvariantCultureIgnoreCase) ||
-                        httpRequestHeader.Equals(SpecVersionHttpHeader2, StringComparison.InvariantCultureIgnoreCase))
+                    // TODO: This is a bit ugly.
+                    var memoryStream = new MemoryStream();

```

Comment:
Need to dispose?

Reply:
Nope. (I'll leave any similar comments later. If you'd prefer that we *do* dispose, we can discuss that in a meeting. Once it's all centralized it would be a trivial change.)